src/frontend/src/utils/redaction.rs

```diff
@@ -0,0 +1,138 @@
+// Copyright 2023 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use risingwave_connector::source::UPSTREAM_SOURCE_KEY;
+use risingwave_sqlparser::ast::*;
+
+use crate::utils::WithOptions;
+
+/// Returns a redacted statement, or None if fails.
+///
+/// It redacts properties in create source/create table/create sink, according to `MarkRedaction::marks`.
+///
+/// TODO #14115: redact table name and column names.
+#[allow(dead_code)]
+pub fn try_redact_statement(stmt: &Statement) -> Option<Statement> {
+    use mark_redaction::MarkRedaction;
+    use risingwave_connector::match_source_name_str;
+    use risingwave_connector::source::SourceProperties;
+
+    let mut stmt = stmt.clone();
+    let sql_options = match &mut stmt {
+        Statement::CreateSource { stmt } => &mut stmt.with_properties.0,
+        Statement::CreateTable { with_options, .. } => with_options,
+        Statement::CreateSink { stmt } => &mut stmt.with_properties.0,
+        _ => {
+            return Some(stmt);
+        }
+    };
+    let Ok(mut with_properties) =
+        WithOptions::try_from(sql_options.as_slice()).map(WithOptions::into_inner)
+    else {
+        return None;
+    };
+    let Some(connector) = with_properties.remove(UPSTREAM_SOURCE_KEY) else {
+        tracing::error!("Must specify 'connector' in WITH clause");
+        return None;
+    };
```

Comment:
What about finding a way to unify this snippet with ...

https://github.com/risingwavelabs/risingwave/blob/5417d18ee47628df4c668e23966604c7fb9c5a86/src/frontend/src/utils/with_options.rs#L208-L246